### PR TITLE
[air/docs] checkpoints

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

Add some documentation about checkpoints.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-air/doc_code/checkpoint_usage.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+from sklearn.model_selection import train_test_split
+
+from ray.data.preprocessors import *
+
+data_raw = load_breast_cancer()
+dataset_df = pd.DataFrame(data_raw["data"], columns=data_raw["feature_names"])
+dataset_df["target"] = data_raw["target"]
+train_df, test_df = train_test_split(dataset_df, test_size=0.3)
+train_dataset = ray.data.from_pandas(train_df)
+valid_dataset = ray.data.from_pandas(test_df)
```
Comment (matthewdeng) [inline doc/source/ray-air/doc_code/checkpoint_usage.py:33]: Should we include this in the doc so the user can run the code directly and play around with the checkpoint themselves?

File: doc/source/ray-air/doc_code/checkpoint_mlflow.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+# Create checkpoint object from path
+checkpoint = Checkpoint.from_directory("model_directory")
+
+# Convert into dictionary
+checkpoint_dict = checkpoint.to_dict()
+
+# This dict can then be passed around, e.g. to a different function
+
+# At some other location, recover checkpoint object from dict
+checkpoint = Checkpoint.from_dict(checkpoint_dict)
```
Comment (matthewdeng) [inline doc/source/ray-air/doc_code/checkpoint_mlflow.py:21]: Similar concern as above, in this example I don't see what the value of having a `Checkpoint` is.

**Review by krfricke — COMMENTED** [review]
Generally I like this addition a lot. I share the main concern of Matt that the "different location" is not clear. We currently take a directory checkpoint and turn it into a dict to pass around. Maybe this is clearer with a dict checkpoint that we put into a directory and then load again from disk - then we could say something like "e.g. in a different script"

File: doc/source/ray-air/checkpoints.rst [hunk 3]
```diff
@@ -0,0 +1,79 @@
+Checkpoints
+===========
+
+Checkpoints are the common format for models that are used across different components of the Ray AI Runtime.
+
+.. image:: images/checkpoint_diagram.png
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/ray-air/checkpoints.rst:6]: done

File: doc/source/ray-air/doc_code/checkpoint_usage.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+# Create checkpoint object from data
+checkpoint = Checkpoint.from_dict(checkpoint_data)
+
+# Save checkpoint to temporary location
+path = checkpoint.to_directory()
+
+# This path can then be passed around, e.g. to a different function
+
+# At some other location, recover Checkpoint object from path
+checkpoint = Checkpoint.from_directory(path)
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/ray-air/doc_code/checkpoint_usage.py]: take a look now?

File: doc/source/ray-air/doc_code/checkpoint_usage.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+from ray import serve
+from ray.serve import PredictorDeployment
+from ray.serve.http_adapters import json_request
+
+
+async def adapter(request: Request):
+    content = await request.json()
+    print(content)
+    return pd.DataFrame.from_dict(content)
```
Comment (krfricke) [inline doc/source/ray-air/doc_code/checkpoint_usage.py:82]: Do we need this? I think this clutters the example

**Review by krfricke — APPROVED** [review]
Looks good to me, one last comment about the Serve example

File: doc/source/ray-air/checkpoints.rst [hunk 6]
```diff
@@ -0,0 +1,84 @@
+Checkpoints
```
Comment (matthewdeng) [inline doc/source/ray-air/checkpoints.rst:1]: Lint/tests failing

**Review by matthewdeng — APPROVED** [review]

File: doc/source/ray-air/doc_code/checkpoint_usage.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+
+model = build_model()
+
+checkpoint = to_air_checkpoint(model)
+# __checkpoint_quick_end__
+
+
+# __use_trainer_checkpoint_start__
+import ray
```
Comment (richardliaw) [inline doc/source/ray-air/doc_code/checkpoint_usage.py:26]:
```suggestion
import pandas as pd
import ray
```

File: doc/source/ray-air/doc_code/checkpoint_mlflow.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+# ... e.g. train model with clf.fit()
+# Save model using MLflow
+mlflow.sklearn.save_model(clf, "model_directory")
+
+# Create checkpoint object from path
+checkpoint = Checkpoint.from_directory("model_directory")
+
+# Write it to some other directory
+checkpoint.to_directory("other_directory")
+# You can also use `checkpoint.to_uri/from_uri` to
```
Comment (pcmoritz) [inline doc/source/ray-air/doc_code/checkpoint_mlflow.py:20]: I'm a bit surprised that `.to_directory(dir)` is not just `.to_uri("file://" + dir)`. Why have two methods?

File: doc/source/ray-air/doc_code/checkpoint_usage.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            tf.keras.layers.InputLayer(input_shape=(1,)),
+            tf.keras.layers.Dense(1),
+        ]
+    )
+    return model
+
+
+model = build_model()
+
+checkpoint = to_air_checkpoint(model)
```
Comment (pcmoritz) [inline doc/source/ray-air/doc_code/checkpoint_usage.py:21]: Feels a bit clunky, why not `checkpoint = model.to_checkpoint()`?

Comment (richardliaw) [reply doc/source/ray-air/doc_code/checkpoint_usage.py:21]: there is no "model" class in Air.

File: doc/source/ray-air/doc_code/checkpoint_usage.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+import pandas as pd
+
+from ray import serve
+from ray.serve import PredictorDeployment
+from ray.serve.http_adapters import json_request
+
+
+async def adapter(request: Request):
+    content = await request.json()
+    print(content)
```
Comment (pcmoritz) [inline doc/source/ray-air/doc_code/checkpoint_usage.py:85]: spurious print?

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**